backend/apps/rag/batch_utils.py

```diff
@@ -0,0 +1,254 @@
+    for text in texts:
+        tokens: int = text2tokens(text)
+        if current_tokens + tokens > token_limit:
+            yield (
+                batch_idx,
+                b"\n".join(
+                    _request_iterator(
+                        current_request, model, uuid, batch_idx, chunk_size
+                    )
+                ),
+            )
+            batch_idx += 1
+            current_request = [text]
+            current_tokens = tokens
+        else:
+            current_request.append(text)
+            current_tokens += tokens
+    if current_request:
+        yield (
+            batch_idx,
+            b"\n".join(
+                _request_iterator(current_request, model, uuid, batch_idx, chunk_size)
+            ),
+        )
+
+
+def _request_async_embedding(
+    model: str,
+    texts: list[str],
+    key: str,
+    url: str = "https://api.openai.com/v1",
+    completion_window: Literal["24h"] = "24h",
+    wait_interval: int = 60,
+    token_limit: int = 1_000_000,  # Basic TPM for Tier 1, assuming one batch is processed per minute
+) -> list[list[float]]:
+    uuid: str = uuid4().hex
+    client: OpenAI = OpenAI(api_key=key, base_url=url)
+    return [
+        embedding
+        for batch_idx, batch in _batch_iterator(texts, model, uuid, token_limit)
+        for embedding in _process_one_batch(
+            client, uuid, batch_idx, batch, completion_window, wait_interval
+        )
+    ]
+
+
+@retry(stop=stop_after_attempt(5), wait=wait_exponential(multiplier=1, min=4, max=10))
+def _request_sync_embedding(url, key, texts, model):
+    log.info("Using single processing for OpenAI embeddings")
+    r = requests.post(
+        f"{url}/embeddings",
+        headers={
+            "Content-Type": "application/json",
+            "Authorization": f"Bearer {key}",
+        },
+        json={"input": texts, "model": model},
+    )
+    r.raise_for_status()
+    data = r.json()
+    if "data" in data:
+        return [elem["embedding"] for elem in data["data"]]
+    else:
+        log.error(data)
+        raise Exception("Something went wrong :/")
+
+
+def generate_openai_batch_embeddings(
+    model: str, texts: list[str], key: str, url: str = "https://api.openai.com/v1"
+) -> Optional[list[list[float]]]:
+    try:
+        text_length: int = sum((len(text) for text in texts))
+        log.info(
+            f"len(texts): {len(texts)} / sum(len(text) for text in texts): {text_length}"
+        )
+        if text_length < 500_000:
+            log.info("Using single processing for OpenAI embeddings")
+            func = _request_sync_embedding
+            return _request_sync_embedding(url, key, texts, model)
+        else:
+            log.info("Using batch processing for OpenAI embeddings")
+            func = _request_async_embedding
+        return func(model, texts, key, url)
+
+    except Exception as e:
+        log.exception(e)
+        return None
+
+
+if __name__ == "__main__":
```

Comment:
tests should be in a separate test file.

Reply:
@tjbck Ok then I should mock the openai client to test the module. Working on it.